lib/libvfio-user.c

```diff
@@ -927,126 +901,216 @@ device_reset(vfu_ctx_t *vfu_ctx, vfu_reset_type_t reason)
+                // "supported" if said support is just an empty callback?
+                //
+                // We don't need to return RUNNING or ERROR since they are always
+                // supported.
+                mig->flags = VFIO_MIGRATION_STOP_COPY | VFIO_MIGRATION_PRE_COPY;
+
+                ret = 0;
+            } else {
+                assert(feature == VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE);
+
+                struct vfio_user_device_feature_mig_state *state =
+                    (void*)res->data;
+                
+                state->device_state = migration_get_state(vfu_ctx);
+
+                ret = 0;
+            }
+        } else if (is_dma_feature(feature)) {
+            struct vfio_user_device_feature_dma_logging_report *rep =
+                (void*)req->data;
+
+            ssize_t bitmap_size = get_bitmap_size(rep->length, rep->page_size);
+
+            msg->out.iov.iov_len = sizeof(struct vfio_user_device_feature)
+                + sizeof(struct vfio_user_device_feature_dma_logging_report)
+                + bitmap_size;
+            msg->out.iov.iov_base = calloc(1, msg->out.iov.iov_len);
+
+            if (msg->out.iov.iov_base == NULL) {
+                return ERROR_INT(ENOMEM);
+            }
+
+            memcpy(msg->out.iov.iov_base, msg->in.iov.iov_base,
+                   sizeof(struct vfio_user_device_feature) +
+                   sizeof(struct vfio_user_device_feature_dma_logging_report));
+
+            struct vfio_user_device_feature *res = msg->out.iov.iov_base;
+
+            res->argsz = msg->out.iov.iov_len;
+
+            dma_controller_t *dma = vfu_ctx->dma;
+
+            assert(dma != NULL);
+
+            char* bitmap = (char*) msg->out.iov.iov_base
+                + sizeof(struct vfio_user_device_feature)
+                + sizeof(struct vfio_user_device_feature_dma_logging_report);
+
+            ret = dma_controller_dirty_page_get(dma,
+                                                (vfu_dma_addr_t) rep->iova,
+                                                rep->length,
+                                                rep->page_size,
+                                                bitmap_size,
+                                                bitmap);
+
+            if (ret < 0) {
+                msg->out.iov.iov_len = 0;
+            }
+        } else {
+            return ERROR_INT(EINVAL);
+        }
+    } else if (req->flags & VFIO_DEVICE_FEATURE_SET) {
+        msg->out.iov.iov_base = malloc(msg->in.iov.iov_len);
+        msg->out.iov.iov_len = msg->in.iov.iov_len;
+
+        if (msg->out.iov.iov_base == NULL) {
+            return ERROR_INT(ENOMEM);
+        }
+
+        memcpy(msg->out.iov.iov_base, msg->in.iov.iov_base,
+               msg->out.iov.iov_len);
+
+        struct vfio_user_device_feature *res = msg->out.iov.iov_base;
+
+        if (is_migration_feature(feature)) {
+            assert(feature == VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE);
+            
```

Comment:
Turn on whitespace highlighting in whatever editor you use - there's some whitespace noise in various places.